Samples/AudioCreation/cs/AudioCreation/Scenario7_RecordPlayback.xaml.cs

```diff
@@ -0,0 +1,570 @@
+                }
+
+                fixed (byte* buf = byteBuffer)
+                {
+                    for (int i = 0; i < capacityInBytes; i++)
+                    {
+                        byte b = dataInBytes[i + bufferStart];
+                        buf[i + maxIndex] = b;
+
+                        if (b != 0)
+                        {
+                            nonZeroSampleCount++;
+                        }
+                    }
+                }
+                maxIndex += capacityInBytes;
+            }
+
+            audioFrameCount++;
+        }
+
+        private void UpdateStatusWhileRecording()
+        {
+            uint max = maxIndex >> 3;
+            uint curr = currentIndex >> 3;
+            int frameCount = audioFrameCount;
+            // may have stopped recording while this async notification was in flight
+            if (isRecording)
+            {
+                rootPage.NotifyUser($"recording: max samples {max}, frame count {frameCount}, last cap {lastCapacityEncountered}, max cap {maxCapacityEncountered}", NotifyType.StatusMessage);
+            }
+        }
+
+        private void FrameInputNode_QuantumStarted(AudioFrameInputNode sender, FrameInputNodeQuantumStartedEventArgs args)
+        {
+            uint requiredSamples = (uint)args.RequiredSamples;
+
+            if (requestStartPlaying)
+            {
+                Debug.Assert(!isRecording);
+                Debug.Assert(!requestStartRecording);
+                Debug.Assert(!isPlaying);
+
+                isPlaying = true;
+                requestStartPlaying = false;
+                currentIndex = 0;
+            }
+
+            if (isPlaying)
+            {
+                Debug.Assert(!isRecording);
+                Debug.Assert(!requestStartPlaying);
+                Debug.Assert(!requestStartRecording);
+
+                HandleOutgoingAudio(requiredSamples);
+
+                /*
+                int currentStatusMsec = DateTime.Now.Millisecond;
+                // update status every tenth of a second
+                if (lastStatusMsec % 100 > currentStatusMsec % 100)
+                {
+                    CoreApplication.MainView.CoreWindow.Dispatcher.TryRunAsync(
+                        Windows.UI.Core.CoreDispatcherPriority.Normal,
+                        UpdateStatusWhilePlaying);
+                }
+                lastStatusMsec = currentStatusMsec;
+                */
+            }
+        }
+
+        // We keep a 1MB AudioFrame and reuse it.
+        static AudioFrame s_audioFrame = new Windows.Media.AudioFrame(1024 * 1024);
+        private int s_zeroByteOutgoingFrameCount;
+
+        /// <summary>
+        /// Handle a frame of outgoing audio to the output device.
+        /// </summary>
+        /// <remarks>
+        /// This pushes a megabyte of audio each time, to greatly reduce the number of callbacks.
+        /// 
+        /// We don't need the requiredSamples argument but we don't 
```

Comment:
1 MB of bytes = 131,072 stereo float samples = 2.73 *seconds* at 48Khz sample rate.  (It's amazing how small audio is these days...)

In my testing, requiredSamples is generally around 480, a 10 msec buffer at 48Khz.  I do not want the system to call me back every 10 msec, especially in a C# / .NET app which might be having a GC.  Giving much bigger amounts of data, and deliberately ignoring the number of required samples (which is a minimum), works very well -- the system happily buffers the extra data and plays through it, calling me back only at the end of the full megabyte.

In my earlier system (http://holofunk.com) it was a critical optimization to always give the ASIO library some quite large buffers, and I was very happy to see the same strategy seems to work with AudioGraph.

I will enhance the comments here to make this clearer.